### PR TITLE
feat(logplex source): Introduce Heroku button

### LINKED ISSUE #2887: Heroku Integration / Heroku Button
Many Heroku customers have asked how they could use Vector to ship their Heroku logs. This is especially true for current Timber Hosted customers as many of them are on Heroku. In order to unblock them, we need to provide a simple solution.

## Background

This is not a new problem. We've attempted to solve this problem a couple of times in the past but the goals and context were different:

### Attempt 1 

Our first attempt at solving this prioritized ease of setup, and attempted to move towards a hosted Vector service. The goal was to provide a blanket solution that would reliably transition _all_ Heroku users at various engineering levels. Heroku customers are generally not interested in hosting infrastructure, configuring networks, and setting up HTTP load balancers. They just want to configure a [Heroku log drain]() and be on their way. Unfortunately, we didn't have a way to achieve this for them, but we did have plans to offer a hosted Vector service at some point in the future. So we decided to take the easiest first step towards a hosted Vector service by deploying Vector in an ECS container for each customer that needed it. From the customer's perspective, this would feel like a hosted Vector service. They'd configure a Heroku log drain and be on their way. As we increased the number of customers on this platform we would then take the next steps towards automation, scaling, and formalizing this as a true Vector hosted service. This was never implemented because of attempt 2...

### Attempt 2

The Alloy team needed something quick and easy to unblock customers, and they attempted to create a [Lambda function](https://github.com/timberio/heroku-lambda-cloudwatch) that customers could deploy in their AWS accounts as a serverless solution. Many of the customers were attempting to use AWS CloudWatch, so having them deploy an AWS Lambda function did not feel intrusive. Because this was done quickly, many of the users waiting on attempt 1 decided to use this. This worked ok for getting logs from Heroku to CloudWatch, but it fell short when the customer wanted to do anything with their data, like parse it. It doesn't make sense to invest any more time or money into this function when this is Vector's domain. And that brings me to the solution proposed in this issue...

## Proposal

I propose that we add Heroku to the list of platforms in our [installation section](https://vector.dev/docs/setup/installation/platforms/). The approach would leverage Heroku's [button feature](https://devcenter.heroku.com/articles/heroku-button), where a user could click the Heroku button and deploy Vector as a dyno in their own Heroku account.

## Requirements

- [ ] Create a Heroku button that deploys Vector into their Heroku account.
- [ ] It should be preconfigured with the `heroku_logplex` source.
- [ ] It should be preconfigured with parsers for the [various Heroku logs](https://devcenter.heroku.com/articles/logging) (this will need fleshing out).
- [ ] It should be preconfigured with the `aws_cloudwatch_logs` sink, or somehow easily configurable for this sink.
- [ ] It should be required that users provide AWS credentials if they are using AWS sinks. This is probably best achieved with environment variables since this is Heroku's means of configuring services.
- [ ] The HTTP URL should be readily available so that users can configure log drains for their other Heroku services to point to the Vector service.

## Open Questions

1. We need to audit and list all of the various log formats that Heroku includes so that we can include parsers. Most of them are in logfmt, you can see a list of [error logs here](https://devcenter.heroku.com/articles/error-codes).
2. Should we also address other common formats, such as Rails. Or should we require users to simplify configure their app to log JSON. I vote for the latter.
3. How can we make it easy for users to configure Vector upon deployment? #1966 proposes a simple pipline syntax that is perfect for scenarios like this. This syntax could be provided as an environment variable.

### LINKED ISSUE #2922: Heroku/HTTP: Sources lack HTTPS User/Password configuration knobs
The Heroku docs display on https://devcenter.heroku.com/articles/log-drains#https-drains a user/pass combination being used.

Also note:

> Security considerations
>
> HTTPS drains support transport-level encryption using the HTTPS protocol, and authentication using HTTP Basic Authentication. However, Logplex does not currently validate the server certificate during TLS handshake.
>
> Syslog drains support transport-level encryption, including certificate and hostname verification during TLS handshake. However, Logplex does not support authentication for syslog drains. Therefore, it is possible for third parties to send log messages to your server if they know its address. As a security measure, you can filter out messages with unknown (or missing) drain tokens.
>
>TLS Syslog drains fully support modern TLS, such as TLS 1.2 and TLS 1.1. Older TLS (v1.0) and all versions of SSL (v1,2,3) are not >supported. Modern Galois/Counter Mode (GCM) based ciphers are preferred, however older Cipher Block Chaining (CBC) is still >supported. However, Logplex uses only 128-bit AES variants of these ciphers. 

Given that `app.json` files have a built in way to define secures (https://devcenter.heroku.com/articles/app-json-schema#env) this means user/pass is the easiest and seemingly best option for users wishing to secure a logplex source.

Unfortunately, our docs do not specify this as a valid option: https://vector.dev/docs/reference/sources/logplex/#configuration

### PR DESCRIPTION
**Pull request by Hoverbear**
Fixes #2887 

Hard blocked by: #2922 
Related: #2921 #2919 

Adds Heroku Deploy button for Vector. 

Some notes:

* It takes a URL and auto refreshes from that URL each boot. Users can opt out by provisioning a `./vector.toml`.
* It pulls the newest stable musl build of vector, if it doesn't exist. This means users always get the new stable vector, and it doesn't update unless the node is take down and re-provisioned.
* It relies on HTTPS Basic Auth to secure the source, (Not implemented in the buildpack yet!) but it's blocked by #2922 .

Further work:

* FIx #2922
* Add HTTPS basic auth to the vector template in the buildpack. (See https://devcenter.heroku.com/articles/log-drains#https-drains for auth)
* The address needs to be changed to point to master not this branch
* @binarylogic level docs pass

### CODE REVIEW

**Comment by Hoverbear** [comment]
@binarylogic Let me know if you want me to get this ready for a review/merge again. :)

File: app.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            "required": false
+        },
+        "AWS_CLOUDWATCH_STREAM_NAME": {
+            "description": "An AWS CloudWatch stream name (if needed, for sinking to CloudWatch. Vector can create this for you.)",
+            "required": false
+        },
+        "AWS_CLOUDWATCH_GROUP_NAME": {
+            "description": "An AWS CloudWatch stream group name (if needed, for sinking to CloudWatch. You need to create this yourself.)",
+            "required": false
+        },
```
Comment (binarylogic) [inline app.json:34]: @Hoverbear are these necessary given that we no longer prefer Cloudwatch? This is really the big thing for me. I want to make sure deploying Vector as a Heroku button is not restricted to AWS or Cloudwatch.

Comment (Hoverbear) [reply app.json:34]: We can totally yank them! :) Where would you like the logs to go then?

Comment (binarylogic) [reply app.json:34]: Couldn't we base it entirely on the remote Vector config?

Comment (Hoverbear) [reply app.json:34]: We could! Yeah! Okay! I'll refactor this :)

**Comment by jamtur01** [comment]
@binarylogic How do you want to proceed here?

**Comment by binarylogic** [comment]
If the remaining work is <= 3 then let's finish it. I just want to make this generic and rip out any AWS/CloudWatch assumptions.

**Comment by Hoverbear** [comment]
https://github.com/timberio/vector/issues/2922 is required, I think it's around a 3 unless there is something I'm not seeing. I removed cloudwatch assumptions.

**Comment by jszwedko** [comment]
If we are just letting people specify arbitrary vector configs, it doesn't seem like this would be blocked by https://github.com/timberio/vector/issues/2922 . Specifically using a Heroku deployed vector to consume logplex logs would be, but not generally deploying vector onto Heroku.

**Comment by Hoverbear** [comment]
Indeed, this is more of a soft-block, but I would be very uncomfortable with someone accidentally leaving their logging router open to the public because we lacked a feature. :(

**Comment by jszwedko** [comment]
True, I'm thinking that they could deploy it using sources that behave like consumers, like the `kafka` source, though.

We should definitely add basic auth support too.

**Comment by Hoverbear** [comment]
Some notes:

* Since we no longer have a default to cloudwatch I moved this to a blackhole.
* Since we have no reasonable default I felt it was required to write more docs.
* Added basic auth support.
* Kept remote URL support but no longer make it recommended. It's preferred if users push to the repo.

This is good for a final review. @jamtur01

**Comment by Hoverbear** [comment]
Oh! Last note, this should wait till 0.11 is out.

**Comment by jamtur01** [comment]
@Hoverbear Now 0.11 is out - can this proceed?

**Comment by Hoverbear** [comment]
Heck yeah!

**Pull request closed without merging**